ext/openssl/openssl_pwhash.c

```diff
@@ -0,0 +1,401 @@
+		&threads);
+	*p++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_ARGON2_LANES,
+		&threads);
+	*p++= OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_ITER,
+		&iterlimit);
+	*p++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_ARGON2_MEMCOST,
+		&memlimit);
+	*p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT,
+		(void *)salt, salt_len);
+	*p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_PASSWORD,
+		(void *)pass, pass_len);
+	*p++ = OSSL_PARAM_construct_end();
+
+	if ((kdf = EVP_KDF_fetch(ctx, algo, NULL)) == NULL) {
+		goto fail;
+	}
+	if ((kctx = EVP_KDF_CTX_new(kdf)) == NULL) {
+		goto fail;
+	}
+	if (EVP_KDF_derive(kctx, hash, hash_len, params) != 1) {
+		zend_value_error("Unexpected failure hashing password");
+		goto fail;
+	}
+
+	ret = true;
+
+fail:
+	EVP_KDF_free(kdf);
+	EVP_KDF_CTX_free(kctx);
+
+	if (threads > 1) {
+		OSSL_set_max_threads(ctx, 0);
+		OSSL_LIB_CTX_free(ctx);
+	}
+	return ret;
+}
+
+static zend_string *php_openssl_argon2_hash(const zend_string *password, zend_array *options, const char *algo)
+{
+	uint32_t iterlimit, memlimit, threads, version = PHP_OPENSSL_ARGON_VERSION;
+	zend_string *digest = NULL, *salt64 = NULL, *hash64 = NULL;
+	unsigned char hash[PHP_OPENSSL_HASH_SIZE+1], salt[PHP_OPENSSL_SALT_SIZE+1];
+
+	if ((ZSTR_LEN(password) >= UINT32_MAX)) {
+		zend_value_error("Password is too long");
+		return NULL;
+	}
+	if (get_options(options, &memlimit, &iterlimit, &threads) == FAILURE) {
+		return NULL;
+	}
+	if (RAND_bytes(salt, PHP_OPENSSL_SALT_SIZE) <= 0) {
+		return NULL;
+	}
+
+	if (!php_openssl_argon2_compute_hash(algo, version, memlimit, iterlimit, threads,
+			ZSTR_VAL(password), ZSTR_LEN(password),	salt, PHP_OPENSSL_SALT_SIZE, hash, PHP_OPENSSL_HASH_SIZE)) {
+		return NULL;
+	}
+
+	hash64 = php_base64_encode_ex(hash, PHP_OPENSSL_HASH_SIZE, PHP_BASE64_NO_PADDING);
+
+	salt64 = php_base64_encode_ex(salt, PHP_OPENSSL_SALT_SIZE, PHP_BASE64_NO_PADDING);
+
+	digest = zend_string_alloc(PHP_OPENSSL_DIGEST_SIZE, 0);
+	ZSTR_LEN(digest) = snprintf(ZSTR_VAL(digest), ZSTR_LEN(digest), "$%s$v=%d$m=%u,t=%u,p=%u$%s$%s",
+		algo, version, memlimit, iterlimit, threads, ZSTR_VAL(salt64), ZSTR_VAL(hash64));
+
+	zend_string_release(salt64);
+	zend_string_release(hash64);
+
+	return digest;
+}
+
+static int php_openssl_argon2_extract(
+	const zend_string *digest, uint32_t *version, uint32_t *memlimit, uint32_t *iterlimit,
+	uint32_t *threads, zend_string **salt, zend_string **hash)
+{
+	const char *p;
+	char *hash64, *salt64;
+
+	if (!digest || (ZSTR_LEN(digest) < sizeof("$argon2id$"))) {
+		return FAILURE;
+	}
+	p = ZSTR_VAL(digest);
+	if (!memcmp(p, "$argon2i$", strlen("$argon2i$"))) {
+		p += strlen("$argon2i$");
+	} else if (!memcmp(p, "$argon2id$", strlen("$argon2id$"))) {
+		p += strlen("$argon2id$");
+	} else {
+		return FAILURE;
+	}
+	if (sscanf(p, "v=%" PRIu32 "$m=%" PRIu32 ",t=%" PRIu32 ",p=%" PRIu32,
+		   version, memlimit, iterlimit, threads) != 4) {
```

Comment:
CS: NIT: I prefer not to mix tabs and spaces in indent - I usually use two tabs indent for continous indent but one is also fine.

Reply:
done